src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -93,25 +94,55 @@ public void RefreshCompilation(DocumentUri documentUri)
             // TODO: This may cause race condition if the user is modifying the file at the same time
             // need to make a shallow copy so it counts as a different file even though all the content is identical
             // this was the easiest way to force the compilation to be regenerated
-            var shallowCopy = new BicepFile(compilationContext.Compilation.SourceFileGrouping.EntryPoint);
+            var shallowCopy = compilationContext.Compilation.SourceFileGrouping.EntryPoint.ShallowClone();
             UpsertCompilationInternal(documentUri, null, shallowCopy);
         }
 
         public void UpsertCompilation(DocumentUri documentUri, int? version, string fileContents, string? languageId = null, bool triggeredByFileOpenEvent = false)
         {
             if (this.ShouldUpsertCompilation(documentUri, languageId))
             {
-                var newFile = SourceFileFactory.CreateSourceFile(documentUri.ToUri(), fileContents);
+                var newFile = CreateSourceFile(documentUri, fileContents, languageId);
+
                 UpsertCompilationInternal(documentUri, version, newFile, triggeredByFileOpenEvent: triggeredByFileOpenEvent);
             }
         }
 
+        private ISourceFile CreateSourceFile(DocumentUri documentUri, string fileContents, string? languageId)
```

Comment:
Can't we be precise here, rather than file extension based? `UpsertCompilation` is only invoked in 2 places - on open (with langid) and on update (without langid, but the `activeContexts` should contain an entry)

Reply:
I agree that this is the right thing because the client would be deciding how file extensions are mapped to language IDs instead of the language server, but there are complications:

1. Fatal exception on file open for named files (not untitled) prevents the file from being added to activeContexts. Subsequent edits would fail to create source file and the user would be stuck until they close the file (assuming that they don't exhaust the LSP retry limit - haven't tested). 

2. ARM templates still have to be loaded by file extension - not a big deal though.

I think number 1 can be fixed, but I don't think it's worth it right now. (Easiest thing is probably to create some sort of broken compilation and semantic model that always returns the one fatal diagnostic.) I'm going to leave as-is.

Reply:
Actually if we could track just the language ID for a file regardless if it opens successfully or not, then we'd be ok.

Reply:
Btw, I'm going to try this out in this PR.

Reply:
Seems like it's working. I also made the `IBicepCompilationManager` contract more explicit. I still need to test template specs and registry modules, though.

Reply:
Verified.